pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/sink/PulsarSink.java

```diff
@@ -400,7 +415,16 @@ public void close() throws Exception {
         ConsumerConfig consumerConfig = new ConsumerConfig();
         consumerConfig.setSchemaProperties(pulsarSinkConfig.getSchemaProperties());
         if (!StringUtils.isEmpty(pulsarSinkConfig.getSchemaType())) {
-            consumerConfig.setSchemaType(pulsarSinkConfig.getSchemaType());
+            if (GenericRecord.class.isAssignableFrom(typeArg)) {
```

Comment:
@sijie initially you pointed out that working here in PulsarSink is not the right way, but we should only work on TopicSchema
https://github.com/apache/pulsar/pull/9481#discussion_r570723440

In fact I believe that in my PR #9481 I took the right way, driven by your suggestions.

I believe that this change is not enough in order to support by needs.
 
 BTW if the integration test I added to #9481 works with this patch then we can converge to a good  solution.
 My goal is to get that usecase work, in the best way for the project for the mid/long term

Reply:
@eolivelli Yes and no on my original comment. 

My original comment is to make sure we returned the write schema information via TopicSchema. Because we are using `AUTO_CONSUME` in the PulsarSink to indicate `GenericRecord` are published to the Pulsar topic. `AUTO_CONSUME`  can be used by both source and sink. In order to not impact sources, I didn't add the logic in `TopicSchema`. Instead, I add it in PulsarSink to make it more explicit, which results in one line of similar change as your initial change. But it doesn't your original and current implementation is in the right direction.

The main problem of your previous and current implementation on #9481 is you are trying to hijack the existing AVRO implementation to introduce the support of lazy schema initialization. The lazy schema initialization is already implemented as part of multi-schema write support. So you don't need to add such a hack.